src/Development/IDE/Core/Rules.hs

```diff
@@ -497,6 +498,25 @@ typeCheckRule = define $ \TypeCheck file -> do
     -- for files of interest on every keystroke
     typeCheckRuleDefinition hsc pm SkipGenerationOfInterfaceFiles
 
+data GetKnownFiles = GetKnownFiles
+  deriving (Show, Generic, Eq, Ord)
+instance Hashable GetKnownFiles
+instance NFData   GetKnownFiles
+instance Binary   GetKnownFiles
+type instance RuleResult GetKnownFiles = HS.HashSet NormalizedFilePath
+
+knownFilesRule :: Rules ()
+knownFilesRule = defineEarlyCutOffNoFile $ \GetKnownFiles -> do
+  alwaysRerun -- TODO: figure out a way to eliminate this
```

Comment:
I think dropping the `alwaysRerun` is quite easy now that it's backed by a "mutable" collection. Every time the underlying collection gets mutated, make sure to also clear the `GetKnowFiles` key by calling `deleteValue`

This is how it's done for `GetFileExists`, in case it's useful guidance

Reply:
Sure, but an `alwaysRerun` early-cuttoff rule that is just a read of a mutable variable seems quite cheap to me, and also more robust in that it will return the correct result if you accidentally forget to invalidate the key.